internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,532 @@
+	rfc6979ExtraDataV0 = [32]uint8{
+		0xa3, 0xeb, 0x4c, 0x18, 0x2f, 0xae, 0x7e, 0xf4,
+		0xe8, 0x10, 0xc6, 0xee, 0x13, 0xb0, 0xe9, 0x26,
+		0x68, 0x6d, 0x71, 0xe8, 0x7f, 0x39, 0x4f, 0x79,
+		0x9c, 0x00, 0xa5, 0x21, 0x03, 0xcb, 0x4e, 0x17,
+	}
+)
+
+// AdaptorSignature is a signature with auxillary data that commits to a hidden
+// value. When an adaptor signature is combined with a corresponding signature,
+// the hidden value is revealed. Alternatively, when combined with a hidden
+// value, the adaptor reveals the signature.
+//
+// An adaptor signature is created by either doing a public or private key
+// tweak of a valid schnorr signature. A private key tweak can only be done by
+// a party who knows the hidden value, and a public key tweak can be done by
+// a party that only knows the point on the secp256k1 curve derived by the
+// multiplying the hidden value by the generator point.
+//
+// Generally the workflow of using adaptor signatures is the following:
+//  1. Party A randomly selects a hidden value and creates a private key
+//     modified adaptor signature of something for which party B requires
+//     a valid signature.
+//  2. The Party B sees the PublicTweak in the adaptor signature, and creates
+//     a public key tweaked adaptor signature for something that party A
+//     requires a valid signature.
+//  3. Since party A knows the hidden value, they can use the hidden value to
+//     create a valid signature from the public key tweaked adaptor signature.
+//  4. When the valid signature is revealed, by being posted to the blockchain,
+//     party B can recover the tweak and use it to decrypt the private key
+//     tweaked adaptor signature that party A originally sent them.
+type AdaptorSignature struct {
+	r           btcec.FieldVal
+	s           btcec.ModNScalar
+	t           btcec.JacobianPoint
+	pubKeyTweak bool
+}
+
+// Serialize returns a serialized adaptor signature in the following format:
+//
+//	 sig[0:32]  x coordinate of the point R, encoded as a big-endian uint256
+//	 sig[32:64] s, encoded also as big-endian uint256
+//	 sig[64:96] x coordinate of the point T, encoded as a big-endian uint256
+//	 sig[96:128] y coordinate of the point T, encoded as a big-endian uint256
+//	 sig[128] 1 if the adaptor was created with a public key tweak, 0 if it was
+//		created with a private key tweak.
+func (sig *AdaptorSignature) Serialize() []byte {
+	var b [AdaptorSignatureSize]byte
+	sig.r.PutBytesUnchecked(b[0:32])
+	sig.s.PutBytesUnchecked(b[32:64])
+	sig.t.ToAffine()
+	sig.t.X.PutBytesUnchecked(b[64:96])
+	sig.t.Y.PutBytesUnchecked(b[96:128])
+	if sig.pubKeyTweak {
+		b[128] = 1
+	} else {
+		b[128] = 0
+	}
+	return b[:]
+}
+
+func ParseAdaptorSignature(b []byte) (*AdaptorSignature, error) {
+	if len(b) != AdaptorSignatureSize {
+		str := fmt.Sprintf("malformed signature: wrong size: %d", len(b))
+		return nil, errors.New(str)
+	}
+
+	var r secp256k1.FieldVal
+	if overflow := r.SetByteSlice(b[0:32]); overflow {
```

Comment:
Commented on later code, but all of these could similarly be done more efficiently with casts.